### PR TITLE
[ACNA-1357] Optimize deploy to skip posting unchanged actions

### PR DESCRIPTION
**Pull request by florind12**
While `aio app run` on file change, it should redeploy only actions for the modified files. 
It should as well optimize the deployment process by skipping unchanged actions. 

## Description

The build and the deploy steps were altered. 
The build process now includes an extra preparation step. This step is meant to collect data about the actions which are about to be built. 
On the build process, each action from the previous step is validated by a log file that contains/will contain an object with the previously built actions.

Eg: `{"action":"1896e19a5bc531a678a9"}. `

If the action was not built before, aka wasn't found in the file, it will be processed, and added to the file. This indicates that this build is valid for deployment. 

The deployment process triggered by the `aio app run` utilize a new parameter ( which indicated that it should validate the actions before deploying. This validation implies. cutting out the actions from the manifest file.

The already built zip files generated on the buildProcess described above are being used to validate actions from the manifest. By this, we leave out the actions which were not built and go on deploying only the built actions. 

In summary, from now on, it will deploy only actions which were built and will skip building actions that were built before based on the `contenthash` (generated by webpack) or the file modification time if it's a folder not handled by the webpack.

## Related Issue
https://github.com/adobe/aio-cli-plugin-app/issues/474

## How Has This Been Tested?
Manual testing.

- **Deploying optimization**: 
**Steps**: 
On a fresh `aio` generated project. 

First Run: `> aio app run`
_Expected_: All actions should be deployed. 

Second Run: `> aio app run` 
_Expected_: No actions should be deployed. 

Edit action file. 
Third Run: `> aio app run` 
_Expected_: Modified action should be deployed. 
Note: 'aio app run' needs to be closed before running a new instance.  

- **Skip deploying unchanged actions while watcher.**: 

**Steps**: 
On a fresh `aio` generated project. 

Run: `> aio app run`
_Expected_: All action should be redeployed. 

Considering there are at least two actions registered for the project, change 1 action file.
Expected: It should redeploy only the action for the modified file. 

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] I have signed the [Adobe Open Source CLA](http://opensource.adobe.com/cla.html).
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by florind12** [comment]
Changes in dependency tree requires restart. 
Where is the best to put this info up ?

**Comment by florind12** [comment]
Tested this with symlink for the actions folder. Everything seems to be working as expected.

File: test/build.actions.test.js [hunk 1]
```diff
@@ -579,6 +582,23 @@ test('should not fail if extra package does not have actions', async () => {
     path.normalize('/dist/actions/action-zip.zip'))
 })
 
+test('should skip check for action built before while action filter', async () => {
```
Comment (shazron) [inline test/build.actions.test.js:585]: better description for the test needed. maybe `should always zip files because the action because skipCheck=true`

**Review by shazron — CHANGES_REQUESTED** [review]

File: src/utils.js [hunk 2]
```diff
@@ -1994,6 +2007,57 @@ function activationLogBanner (logFunc, activation, activationLogs) {
   }
 }
 
+/**
+ * Returns the parsed object.
+ *
+ * @param {string} stringData stringData you want to be parsed
+ * @returns {object} parsedData or empty object
+ */
+function tryParseString (stringData) {
```
Comment (shazron) [inline src/utils.js:2016]: there is already utils.safeParse, can it not be used?

**Review by shazron — APPROVED** [review]

**Comment by shazron** [comment]
Tested with https://github.com/adobe/aio-cli-plugin-app/pull/497, works ok now.

**Review by sandeep-paliwal — APPROVED** [review]
LGTM

**Pull request merged**